### ISSUE TITLE
reindexArrayWithNumericKeysOtherwiseJavaScriptWillMessWithTheOrder is VERY slow with large nested arrays

### ISSUE THREAD

**Issue by dubcanada** [livewire/livewire#1207]
### Description

I have a large nested array (40 items nested inside each other along with properties, basically a file tree) and doing anything with that takes about 3 seconds each request on local host. After debugging it seems that `reindexArrayWithNumericKeysOtherwiseJavaScriptWillMessWithTheOrder` is extremely slow with large nested arrays.

The uksort part end up looking like

805 x uksort 1 (9.18%) | 348.09ms
805 x uksort 2 (7.44%) | 281.99ms

for the first uksort and for the second uksort. This ends up take about 700ms and each uksort is ran 805 times with the array map being ran 539 times. The `getPublicPropertiesDefinedBySubClass` is ran 5 times.

A large amount of these stats are meaningless to most people, basically every property you add to livewire increases the request time exponentially and adding large (or medium) nested arrays murder the performance.

I think there needs to be a better way of doing `reindexArrayWithNumericKeysOtherwiseJavaScriptWillMessWithTheOrder` that doesn't cause each property you add to decrease performance so drastically.

Anyone have any ideas?

### Steps to reproduce

**Context**
- Livewire version: 1.2
- Laravel version: 7.15
- Browser: Chrome

**Comment by mattharcourt** [comment]
You might try separating the keys differently in `reindexArrayWithNumericKeysOtherwiseJavaScriptWillMessWithTheOrder`. It should be faster if the first uksort is replaced with a loop to separate the numeric keys from the string keys. Maybe replace the two uksorts with something like (not tested):
```php
        $numericKeyed = [];
        $stringKeyed = [];

        foreach ($value as $key => $value) {
            if (is_numeric($key)) {
                $numericKeyed[$key] = $value;
            } else {
                $stringKeyed[$key] = $value;
            }
        }

        uksort($numericKeyed, function ($a, $b) {
            return $a > $b;
        });

        $normalizedData = array_merge($numericKeyed, $stringKeyed);

```

**Comment by jonneroelofs** [comment]
While checking out your issue I think I found another bug in this method (#1257).
I made a PR to fix that bug #1258 
Maybe you could see if that has any impact on your issue? In my solution I combined both uksorts into 1 uksort. So I was hoping it may also help to address your performance issue.

If it doesn't could you provide an/the array (or a similar one)  or code snippet which could be used to benchmark the performance issues?

**Comment by mblarsen** [comment]
Aside: what about renaming `reindexArrayWithNumericKeysOtherwiseJavaScriptWillMessWithTheOrder` to `r63r`

**Comment by calebporzio** [comment]
@jonneroelofs and I did some testing.

We set a component property to an array with 200 items that were all arrays of 200 items themselves.

So 40,000 array entries causes the entire Livewire request (on my machine) to go from ~75ms to ~500ms.

I don't imagine we could do much better than this given the amount of data being processed both on the server and over the wire.

Also, we completely bypassed the uksort and didn't see a big improvement.

For now, I'm going to close this, but feel free to dig deeper and provide more tangible examples and maybe do some digging yourself and re-post the issue.

Thanks!
